### PR TITLE
Ensure method_added does not trigger on a BasicObject singleton class

### PR DESCRIPTION
**Pull request by wied03**
I think this is the best way to test this but I'm open to ideas

### CODE REVIEW

File: core/module/define_method_spec.rb [hunk 1]
```diff
@@ -115,6 +115,23 @@ def test_method
...
+      define_method(:method_added) { |name| method_added_messages << "howdy #{name}" }
+    end
+
+    def_target.class_exec do
+      define_method(:method_added) { |*args| }
+    end
+
+    method_added_messages.should == []
+  end
+
```
Comment (eregon) [inline core/module/define_method_spec.rb]: Could you move it after `it "calls #method_added after the method is added to the Module" do`?

File: core/module/define_method_spec.rb [hunk 2]
```diff
@@ -115,6 +115,23 @@ def test_method
...
+    def_target = class << object; self; end
+
+    method_added_messages = []
+
+    def_target.class_exec do
+      define_method(:method_added) { |name| method_added_messages << "howdy #{name}" }
+    end
+
+    def_target.class_exec do
+      define_method(:method_added) { |*args| }
```
Comment (eregon) [inline core/module/define_method_spec.rb]: Should this define another method? Defining `method_added` to test `method_added` seems quite meta :smiley:

File: core/module/define_method_spec.rb [hunk 3]
```diff
@@ -115,6 +115,23 @@ def test_method
...
 
 describe "Module#define_method" do
+  it "does not call method_added on a BasicObject singleton class" do
+    object = BasicObject
+    def_target = class << object; self; end
+
+    method_added_messages = []
+
+    def_target.class_exec do
+      define_method(:method_added) { |name| method_added_messages << "howdy #{name}" }
```
Comment (eregon) [inline core/module/define_method_spec.rb]: The ScratchPad could be used here for consistency with other specs.

**Comment by eregon** [comment]
> I think this is the best way to test this but I'm open to ideas

I guess you mean that mocks are not available due to how they are implemented in `mspec`?
Looks fine to me and rather easy to understand.

File: core/module/define_method_spec.rb (hunk 3 above)
Comment (elia) [inline core/module/define_method_spec.rb]: Using ScratchPad it maybe can look like:

```rb
ScratchPad.record = []

def BasicObject.method_added(name)
  ScratchPad << name
end

def BasicObject.method_added_to_basic_object
end

ScratchPad.recorded.should == []
```

**Comment by eregon** [comment]
The rule seems to be different, it seems `method_added` is not called for methods added on a `singleton class`, but it is on subclasses of BasicObject (method_added must be defined on the module itself as a singleton method, not an instance_method):
https://gist.github.com/eregon/042337b833c859e4053f

**Comment by eregon** [comment]
Instead, `singleton_method_added` is called when a method is added to a singleton class.

**Comment by eregon** [comment]
See https://github.com/ruby/ruby/blob/ab1a804569237a32fb21bb482476a592919f8789/vm_method.c#L607-L636 for the impl in MRI. I'll write a spec for this.

**Comment by eregon** [comment]
@wied03 @elia 793b0f5 should add specs testing the proper behavior.
Could you test if they catch the expected success/failures in Opal?

File: core/module/define_method_spec.rb (hunk 3 above)
Comment (eregon) [inline core/module/define_method_spec.rb]: Yes, exactly, that's what I meant.

**Comment by elia** [comment]
@eregon :+1: will wait for @wied03 to give the final word

**Comment by eregon** [comment]
BTW this made me discover a funny case with `singleton_method_added`: https://bugs.ruby-lang.org/issues/12131

**Comment by elia** [comment]
Funny indeed, and unexpected, makes rather clear how the implementation works tho :smile:

**Comment by wied03** [comment]
Thanks for cleaning up my mess. Yesterday was a long day.

**Comment by wied03** [comment]
Will check Opal and report back.

**Pull request closed without merging**